core/controllers/admin.py

```diff
@@ -854,3 +854,32 @@ def post(self):
             self.render_json({})
         else:
             raise self.InvalidInputException('This app cannot send emails.')
+
+
+class UpdateUsernameHandler(base.BaseHandler):
+    """Handler for renaming usernames."""
+
+    @acl_decorators.can_access_admin_page
+    def put(self):
+        current_username = self.payload.get('current_username', None)
+        new_username = self.payload.get('new_username', None)
+
+        if new_username is None or current_username is None:
+            raise self.InvalidInputException(
+                'Please ensure that the usernames are not empty.')
+
+        user_id = user_services.get_user_id_from_username(current_username)
+        if user_id is None:
+            raise self.InvalidInputException('User does not exist.')
+
+        if len(new_username) > constants.MAX_USERNAME_LENGTH:
+            raise self.InvalidInputException(
+                'Please make sure that the new username is not longer '
+                'than %s characters.' % constants.MAX_USERNAME_LENGTH)
+        if user_services.is_username_taken(new_username):
+            raise self.InvalidInputException('Username already taken.')
+
+        user_services.set_username(user_id, new_username)
+        user_services.log_username_change(
```

Comment:
is there any specific reason we aren't storing the user_id of the user in the Audit Model?

Reply:
We are storing the user_id of the user(admin) that makes the change. For the user who's username is being changed,  we can easily map their new_username to their user_id since all usernames are unique, so adding their user_id would be redundant.